src/common/cockpitwebserver.c

```diff
@@ -1029,10 +1029,15 @@ on_request_input (GObject *pollable_input,
   gssize count;
 
   length = request->buffer->len;
-  g_byte_array_set_size (request->buffer, length + 4096);
+  /* Read all pending data; with some Gio versions and TLS this is not
+   * level-triggered, i. e. if there is more data available, the callback does
+   * not get called again; thus read up to our allowed maximum size. Plus one
+   * extra byte so that parse_and_process_request() correctly rejects requests
+   * that are > 8 KiB, instead of hanging. */
+  g_byte_array_set_size (request->buffer, length + cockpit_webserver_request_maximum + 1);
 
   count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
-                                                    4096, NULL, &error);
+                                                    cockpit_webserver_request_maximum + 1, NULL, &error);
```

Comment:
you need to subtract `length` here or you'll overrun the buffer

Reply:
No, the g_byte_array_set_size() above reserves length + cockpit_webserver_request_maximum + 1 bytes, so it should be fine. (The previous code did exactly the same, that didn't really change except for renaming the constant)

Reply:
This is true.  But then the above code doesn't make sense: why should we be reserving space in the buffer to potentially read as much as 16k?  We should just fix the buffer at 8k+1, because it can't ever be smaller (given the request size) and should never be bigger (given the limit).

Reply:
Agreed, the current code is unnecessarily complicated. I really want to clean it up, but I prefer doing that in a follow-up (for backporting the minimally intrusive fix).

Reply:
I don't think it would hurt to remove the `+ length` in the call to `g_byte_array_set_size()` and add a `- length` here.  I won't further block you if you want to push back on it, though.

Reply:
@stefwalter just told me that this code is supposed to handle multiple pipelined requests, that's the reason for this structure. It's also why in the previous version we really should have called `process_request()` inside the loop -- it would look for an end of the request, process it, cut it out of the buffer, and let the next  reads just queue it back up. That's the reason behind checking for `cockpit_webserver_request_maximum * 2` and *not* having `- length` here.

So we actually have two different bugs here: (1) that issue with not calling the callback on pending data, and (2) 4 KiB is not enough for some legitimate SSO requests.

I'd like to fix this PR again to drop the ` * 2` change from above, fix the comments, and keep it to an absolute minimum. Then I'll rework #13372 to add a pipelining unit test and then see what still can be simplified.